File: crates/deepsec-processor/src/triage.rs
use crate::agents::{AgentBackend, RevalidateInputFinding, TriageInput};
use crate::errors::ProcessorError;
use deepsec_core::ids::{generate_run_id, now_iso};
use deepsec_core::store::{load_all_file_records, write_file_record};
use deepsec_core::{
    DataRoot, Exploitability, Impact, InvocationMode, ProcessorConfig, RunPhase, RunType,
    Triage, TriagePriority, complete_run, create_run_meta, write_run_meta,
};
use indexmap::IndexMap;
use std::path::PathBuf;

pub struct TriageOptions {
    pub project_id: String,
    pub project_root: PathBuf,
    pub data_root: DataRoot,
    pub backend: std::sync::Arc<dyn AgentBackend>,
    pub filter_prefix: Option<String>,
    pub force: bool,
}

#[derive(Debug, Default, Clone)]
pub struct TriageOutcome {
    pub run_id: String,
    pub triaged: usize,
}

pub async fn run_triage(opts: TriageOptions) -> Result<TriageOutcome, ProcessorError> {
    let run_id = generate_run_id();
    let mut meta = create_run_meta(
        &opts.project_id,
        &run_id,
        opts.project_root.to_string_lossy().as_ref(),
        RunType::Process,
    );
    meta.processor_config = Some(ProcessorConfig {
        agent_type: opts.backend.kind().as_str().into(),
        model: opts.backend.model().into(),
        model_config: IndexMap::new(),
        invocation_mode: Some(InvocationMode::Scan),
        source: None,
    });
    write_run_meta(&opts.data_root, &meta)?;

    let records = load_all_file_records(&opts.data_root, &opts.project_id)
        .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?;
    let mut outcome = TriageOutcome {
        run_id: run_id.clone(),
        ..Default::default()
    };

    for mut rec in records {
        if let Some(prefix) = &opts.filter_prefix {
            if !rec.file_path.starts_with(prefix) {
                continue;
            }
        }
        let mut changed = false;
        let len = rec.findings.len();
        for i in 0..len {
            if !opts.force && rec.findings[i].triage.is_some() {
                continue;
            }
            let f = &rec.findings[i];
            let input = TriageInput {
                file_path: rec.file_path.clone(),
                finding: RevalidateInputFinding {
                    index: i,
                    severity: f.severity,
                    vuln_slug: f.vuln_slug.clone(),
                    title: f.title.clone(),
                    description: f.description.clone(),
                    line_numbers: f.line_numbers.clone(),
                },
            };
            match opts.backend.triage(&input).await {
                Ok((t, _u, _d)) => {
                    let priority = match t.priority.as_str() {
                        "P0" => TriagePriority::P0,
                        "P1" => TriagePriority::P1,
                        "P2" => TriagePriority::P2,
                        _ => TriagePriority::Skip,
                    };
                    let exploit = match t.exploitability.as_str() {
                        "trivial" => Exploitability::Trivial,
                        "moderate" => Exploitability::Moderate,
                        _ => Exploitability::Difficult,
                    };
                    let impact = match t.impact.as_str() {
                        "critical" => Impact::Critical,
                        "high" => Impact::High,
                        "medium" => Impact::Medium,
                        _ => Impact::Low,
                    };
                    rec.findings[i].triage = Some(Triage {
                        priority,
                        exploitability: exploit,
                        impact,
                        reasoning: t.reasoning,
                        triaged_at: now_iso(),
                        model: opts.backend.model().into(),
                    });
                    outcome.triaged += 1;
                    changed = true;
                }
                Err(ProcessorError::Quota(q)) => {
                    tracing::warn!("triage quota exhausted: {} {}", q.backend, q.detail);
                    break;
                }
                Err(e) => {
                    tracing::error!("triage failed for {}: {e}", rec.file_path);
                }
            }
        }
        if changed {
            write_file_record(&opts.data_root, &rec)
                .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?;
        }
    }

    write_run_meta(&opts.data_root, &meta)?;
    complete_run(&opts.data_root, &opts.project_id, &run_id, RunPhase::Done)
        .map_err(|e| ProcessorError::Other(anyhow::Error::new(e)))?;
    Ok(outcome)
}

File: internal/processor/triage.go
package processor

import (
	"context"

	"github.com/noeljackson/deepsec/internal/core"
)

// TriageOptions configures a triage run.
type TriageOptions struct {
	ProjectID    string
	ProjectRoot  string
	DataRoot     core.DataRoot
	Backend      AgentBackend
	ProviderName string
	FilterPrefix string
	Force        bool
}

// TriageOutcome summarizes a triage run.
type TriageOutcome struct {
	RunID   string
	Triaged int
}

// Triage assigns priority/exploitability/impact to findings that don't
// already have a triage field (or every finding, if Force is set).
func Triage(ctx context.Context, opts TriageOptions) (*TriageOutcome, error) {
	runID := core.GenerateRunID()
	meta := core.NewRunMeta(opts.ProjectID, runID, opts.ProjectRoot, core.RunTypeProcess)
	meta.ProcessorConfig = &core.ProcessorConfig{
		AgentType:      opts.ProviderName,
		Model:          opts.Backend.Model(),
		ModelConfig:    map[string]any{},
		InvocationMode: core.InvocationModeScan,
	}
	if err := opts.DataRoot.WriteRunMeta(meta); err != nil {
		return nil, err
	}
	records, err := opts.DataRoot.LoadAllFileRecords(opts.ProjectID)
	if err != nil {
		return nil, err
	}
	out := &TriageOutcome{RunID: runID}
	for _, rec := range records {
		if opts.FilterPrefix != "" && !hasPrefix(rec.FilePath, opts.FilterPrefix) {
			continue
		}
		changed := false
		for i := range rec.Findings {
			if !opts.Force && rec.Findings[i].Triage != nil {
				continue
			}
			f := rec.Findings[i]
			in := &TriageInput{
				FilePath: rec.FilePath,
				Finding: RevalidateInputFinding{
					Index:       i,
					Severity:    f.Severity,
					VulnSlug:    f.VulnSlug,
					Title:       f.Title,
					Description: f.Description,
					LineNumbers: f.LineNumbers,
				},
			}
			t, _, _, err := opts.Backend.Triage(ctx, in)
			if err != nil {
				if IsQuotaErr(err) {
					goto done
				}
				continue
			}
			rec.Findings[i].Triage = &core.Triage{
				Priority:       t.Priority,
				Exploitability: t.Exploitability,
				Impact:         t.Impact,
				Reasoning:      t.Reasoning,
				TriagedAt:      core.NowISO(),
				Model:          opts.Backend.Model(),
			}
			out.Triaged++
			changed = true
		}
		if changed {
			if err := opts.DataRoot.WriteFileRecord(rec); err != nil {
				return nil, err
			}
		}
	}
done:
	if err := opts.DataRoot.WriteRunMeta(meta); err != nil {
		return nil, err
	}
	if _, err := opts.DataRoot.CompleteRun(opts.ProjectID, runID, core.RunPhaseDone); err != nil {
		return nil, err
	}
	return out, nil
}
